src/postprocess/band_structure.jl

```diff
@@ -30,39 +32,17 @@ function high_symmetry_kpath(model; kline_density=40)
                 kpath=[["-½", "½"]], kbranches=[1:length(kcoords)])
     end
 
-    # - Brillouin.jl expects the input direct lattice to be in the conventional lattice
-    #   in the convention of the International Table of Crystallography Vol A (ITA).
-    # - spglib uses this convention for the returned conventional lattice,
-    #   so it can be directly used as input to Brillouin.jl
-    # - The output k-Points and reciprocal lattices will be in the CDML convention.
-    conv_latt = get_spglib_lattice(model; to_primitive=false)
-    sgnum     = spglib_spacegroup_number(model)  # Get ITA space-group number
-    direct_basis   = Bravais.DirectBasis(collect(eachcol(conv_latt)))
-    primitive_latt = Bravais.primitivize(direct_basis, Bravais.centering(sgnum, 3))
-
-    primitive_latt ≈ collect(eachcol(model.lattice)) || @warn(
-        "DFTK's model.lattice and Brillouin's primitive lattice do not agree. " *
-        "The kpath selected to plot the band structure might not be most appropriate."
-    )
-
-    kp     = Brillouin.irrfbz_path(sgnum, direct_basis)
+    # TODO: Wrap this into a function? (Repeated pattern in spglib.jl)
+    spg_positions, spg_numbers, _ = spglib_atoms(model.atoms)
+    cell = Spglib.Cell(model.lattice, spg_positions, spg_numbers)
+
+    kp     = Brillouin.irrfbz_path(cell)
     kinter = Brillouin.interpolate(kp, density=kline_density)
 
     # TODO Need to take care of time-reversal symmetry here!
     #      See https://github.com/JuliaMolSim/DFTK.jl/pull/496/files#r725203554
 
-    # Need to double the points whenever a new path starts
-    # (for temporary compatibility with pymatgen)
-    # TODO Remove this later
-    kcoords = empty(first(kinter.kpaths))
-    for kbranch in kinter.kpaths
-        idcs = findall(k -> any(sum(abs2, k - kcomp) < 1e-5
-                                for kcomp in values(kp.points)), kbranch)
-        @assert length(idcs) ≥ 2
-        idcs = idcs[2:end-1]  # Don't duplicate first and last
-        idcs = sort(append!(idcs, 1:length(kbranch)))
-        append!(kcoords, kbranch[idcs])
-    end
+    kcoords = vcat(kinter.kpaths...)
```

Comment:
I wonder if it is wise to "throw away" the branches here. Would it not simplify things to just keep the list of list of kpoints here and only issue a `vcat( )` when constructing the PlaneWaveBasis? That way you might be able to get rid of the `kdistances_and_ticks` function completely without preventing any of the branch-based tricks in the plotting.

Reply:
Sorry for the delay. I'm afraid getting rid of the kdistances_and_ticks function completely is not possible, because one still needs to compute the kdistances and tick labels.

Also, since `prepare_band_data` accepts `band_data`, it knows the k points only through `basis`, i.e. it cannot directly access `kcoords` from `high_symmetry_kpath`. But I agree that completely discarding branch information is not good, so I added `kbranches` to the output of `high_symmetry_kpath` (which was already the case for 1D).

Reply:
Now I am not sure this was a good idea. `branches` can be easily reconstructed from `basis.kcoords_global`. Keeping `branches` generated by Brillouin makes the code a little simpler, but when it comes to user-defined k paths, it is required for the user to provide `branches` by themselves.

If we just throw away `branches` by Brillouin, then for both Brillouin k path and user-defined k path, `branches` can be automatically calculated in `kdistances_and_ticks`. So, the user does not need to provide `branches`.

Reply:
That's a good point. I see two ways out of this:

- If the user does not supply the branches, everything is on one branch (i.e. what you do now)
- We return Brillouin's `kbranches`, but don't actually use them, i.e. this argument in the returned named tuple is just for information and we re-compute the kbranches as has been done before.

I am tempted to go for the first option, because "custom kpaths" is sort of an advanced feature and then people probably know what they want wrt. the branches. Also to me `prepare_band_data` has always been an internal function (it's not exported for that reason).

But of course @jaemolihm ... if you have a use case which gets harder with this refactoring, than please do speak up ... after all you are doing the work :smile:.